lib/Driver/Job.cpp

```diff
@@ -52,6 +64,29 @@ void CommandOutput::ensureEntry(StringRef PrimaryInputFile,
   }
 }
 
+bool CommandOutput::checkInvariants() const {
+  types::forAllTypes([&](types::ID Type) {
+      size_t numOutputsOfType = 0;
+      for (auto const &I : Inputs) {
+        assert(!I.Primary.empty());
+        auto const *M = DerivedOutputMap.getOutputMapForInput(I.Primary);
+        if (!M)
+          continue;
+        auto const Out = M->find(Type);
+        if (Out == M->end())
+          continue;
+        assert(!Out->second.empty());
+        ++numOutputsOfType;
+      }
+      assert(Type == PrimaryOutputType ||
+             numOutputsOfType == 0 ||
+             numOutputsOfType == 1 ||
+             numOutputsOfType == Inputs.size());
+    });
+  assert(AdditionalOutputTypes.count(PrimaryOutputType) == 0);
+  return true;
```

Comment:
This return value's not actually doing anything. Does it make sense to switch to explicit `NDEBUG` checks, or is the convenience of `assert(checkInvariants())` still worth it?

Reply:
Lazy answer is I can never remember in this codebase whether `!defined(NDEBUG)` is a subset or superset of conditions in which `assert()` is live so I reached for `assert()`. I will switch to `NDEBUG`.